src/main/scala/sparkz/core/network/NodeViewSynchronizer.scala

```diff
@@ -126,13 +126,13 @@ class NodeViewSynchronizer[TX <: Transaction, SI <: SyncInfo, SIS <: SyncInfoMes
       deliveryTracker.setHeld(mod.id)
 
     case SyntacticallyFailedModification(mod, _) =>
-      deliveryTracker.setInvalid(mod.id).foreach(penalizeMisbehavingPeer)
+      deliveryTracker.setInvalid(mod.id).foreach(penalizeAndDisconnectPeer)
```

Comment:
Are we sure that in case the node is syncing and it receives a new forged block (from the tip) it doesn't ban the peer that sent it?

Reply:
I've run full STF migration with these changes - everything passes (some tests fail sporadically, but pass afterwards).
Also, both these exceptions would only appear if the block was chosen to be applied, but failed to pass some checks - making in Invalid.

Reply:
I found this case class returned [here](https://github.com/HorizenOfficial/Sparkz/blob/fix/penalize_invalid_fork/src/main/scala/sparkz/core/NodeViewHolder.scala#L313) if the history.apply() fails. The abstract history apply method in the sdk can fail [here](https://github.com/HorizenOfficial/Sidechains-SDK/blob/dev/sdk/src/main/scala/io/horizen/history/AbstractHistory.scala#L68) if the parent block is not in history. Can't this be happening in the synching phase as Marco mentioned or I'm missing something?
This doesn't apply to the `SemanticallyFailedModification` that is thrown after the apply successfully passed

Reply:
We can safely remove that check. It is there because we get Option type from the storage, so it's a good manner to check and fail with a reasonable error message. But this situation is completely wrong.

The entry point of block application is [here](https://github.com/HorizenOfficial/Sidechains-SDK/blob/0f7e3d1bf5609f6d67e25ab8b2cd2b050c589a52/sdk/src/main/scala/io/horizen/AbstractSidechainNodeViewHolder.scala#L218)
modifiersCache.popCandidate(history()) will only choose blocks with parentId == current best block id, and send them to pmodModify, which in turn invokes history.append().

And this operation is synchronized, so it can't happen that some other thread makes modifications to the local chain at the same time.